### PR TITLE
#792 - Revamp Build Process and Introduce Icon Component

### LINKED ISSUE #493: Huge bundle size
## Describe the bug
The bundle size is huge even developer is using a single component: 
Screenshot 2023-02-11 at 12 07 20

I am using `Import Cost` VSC extension:
> This extension will display inline in the editor the size of the imported package. The extension utilizes webpack in order to detect the imported size.

## Steps to reproduce
1. Install `Import Cost` or `Bundle size` extension in VSC (or do anything else to check package size)
2. Import one of the components from package

## Expected behavior
A library should facilitate tree-shaking, so that if a developer only utilizes a limited number of components from the library, the resulting bundle will not include unused components or code from the library.

## Environment
 - Design system package(s) version: [e.g. @livechat/design-system-react-components@1.0.0-beta.30]

## Additional context
There are a few things that usually break tree-shaking  and cause unused code to be included in the final bundle:
1. Side-effects: Tree-shaking relies on the idea that unused code has no side-effects. If a module has side-effects, it will be included in the final bundle, even if it's not used. To avoid this, you should use the sideEffects property in your package.json file to specify which modules have side-effects.
2. Dynamic imports: Dynamic imports, such as import(), can't be statically analyzed by the bundler, and as a result, all modules imported using this syntax will be included in the final bundle, regardless of whether they're used or not.
3. CommonJS modules: CommonJS is an older module format that's not compatible with tree-shaking. If you're using CommonJS modules, the bundler will include all exports from the module, even if they're not used.
4. Namespace imports: Namespace imports, such as import * as myModule from 'my-module', can't be tree-shaken because the bundler can't determine which exports are used.
5. Unoptimized code: If your code is not optimized (such as having many nested if statements), the bundler may not be able to determine which parts of the code are unused, and as a result, more code may be included in the final bundle.

### LINKED ISSUE #792: New build process for design-system-icons library
### Description

Changing the structure and build process of the design-system-icons library to improve the future developer experience and build size of this library.
The library will now export its own `Icon` component, with a list of available icons integrated into its own props. These lists will be available as enums for corresponding icon sets - for now, `Material` and `Tabler`.
The goal is to automate the building process for icons in enums based on available SVG files in corresponding asset folders.

### PR DESCRIPTION
**Pull request by VadymBezpalko**
Resolves: #493, #792 

## Description
This pull request encompasses two major changes aimed at enhancing the library's functionality and ease of use:

### Build Process Overhaul:

The library's build process has been thoroughly revamped and now leverages the `vite` builder for bundling JavaScript code and generating TypeScript declarations. Additionally, it dynamically generates a TypeScript file named `icons.ts`, which houses lazy-loading imports of SVG icons located in the `assets` folder. These icons are organized within subfolders representing different icon sets, allowing for a structured and efficient icon management system.

### New Icon Component:

```
import { Icon } from '@livechat/design-system-icons';

<Icon name="TextLogo" />
```

A new Icon component has been introduced to provide a more flexible and easy-to-use way of rendering icons within the library. This component comes with the following notable features:

- `set` Prop: Allows choosing between different icon sets. Currently, it supports `tabler` (which is the default) and `material` icon sets.

- `name` Prop: Enables specifying the name of the icon to be rendered. Utilizes TypeScript's discriminating union functionality to ensure that only valid icon names corresponding to the selected icon set are allowed, which helps in catching potential errors at compile-time.

### Bundle size

The Build Process Recamp improves the build sizes of icons and component libraries.

`design-system-icons`
before changes:
image

after changes:
image

`design-system-react-components`
before changes:
ds beta10 info

after changes:
ds remake2 info

## Storybook

https://feature-792--613a8e945a5665003a05113b.chromatic.com

## Checklist

**Obligatory:**

- [x] Self review (use this as your final check for proposed changes before requesting the review)
- [x] Add reviewers (`livechat/design-system`)
- [x] Add correct label
- [x] Assign pull request with the correct issue

### CODE REVIEW

**Review by mreszke — APPROVED** [review]
🔥

File: README.md [hunk 1]
```diff
@@ -40,13 +40,12 @@ You can import components directly from the npm package:
...
-In case of icons there is a separate package to be used in conjunction:
+In case of icons there is a separate package to be used:
 
 ```jsx
-import { Edit } from '@livechat/design-system-icons/react/tabler';
-import { Icon } from '@livechat/design-system-react-components';
+import { Icon } from '@livechat/design-system-icons';
 
-<Icon source={Edit} kind="primary" />;
+<Icon set="tabler" name="Edit" kind="primary" />;
```
Comment (sgraczyk) [inline README.md:48]: Consider making `tabler` the default value for `set`

File: packages/example-react/package.json [hunk 2]
```diff
@@ -1,7 +1,7 @@
 {
   "name": "example-react",
   "private": true,
-  "version": "1.0.0-beta.5",
+  "version": "1.0.0-icons-remake.1",
```
Comment (sgraczyk) [inline packages/example-react/package.json:4]: I'm not sure how to properly release this change as it seems to introduce breaking changes.
We should either release it as alpha, consider new package, or deliver the change in steps, allowing both deprecated form and the new one.

File: packages/example-react/src/App.tsx [hunk 3]
```diff
@@ -1,11 +1,11 @@
...
-import { Badge, Icon } from '@livechat/design-system-react-components';
+import { Icon } from '@livechat/design-system-icons';
+import { Badge } from '@livechat/design-system-react-components';
 
 function App() {
   return (
     <div>
       Hello world <Badge kind="secondary">my badge</Badge>
-      <Icon source={Add} />
+      <Icon name="WoocommerceColored" />
```
Comment (sgraczyk) [inline packages/example-react/src/App.tsx:8]: I'm not suer if it's a good example icon

File: packages/icons/index.html [hunk 4]
```diff
@@ -0,0 +1,13 @@
+<!doctype html>
+<html lang="en">
+  <head>
+    <meta charset="UTF-8" />
+    <link rel="icon" type="image/svg+xml" href="/vite.svg" />
+    <meta name="viewport" content="width=device-width, initial-scale=1.0" />
+    <title>Vite + React + TS</title>
```
Comment (sgraczyk) [inline packages/icons/index.html:7]: Please improve the title.

File: packages/icons/src/App.tsx [hunk 5]
```diff
@@ -0,0 +1,12 @@
+import { Icon } from '../';
+
+function App() {
+  return (
+    <>
+      <Icon name="WoocommerceColored" size="xlarge" customColor="#0000ff" />
+      <div>Hello icon components library!</div>
```
Comment (sgraczyk) [inline packages/icons/src/App.tsx:7]: I would improve this example if possible. It should help testing any changes to this library.

**Review by sgraczyk — APPROVED** [review]

File: README.md (hunk 1 above)
Comment (sgraczyk) [reply README.md:48]: Ok, I see it is handled.

**Review by JoannaSikora — APPROVED** [review]

**Review by marcinsawicki — APPROVED** [review]

**Comment by VadymBezpalko** [comment]
Closing this PR in favor of the better solution: https://github.com/livechat/design-system/pull/815

**Pull request closed without merging**